### PR TITLE
Add Nx.Defn.compile/3, closes #823

### LINKED ISSUE #823: Add Nx.Defn.compile

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: nx/lib/nx/defn/compiler.ex [hunk 1]
```diff
@@ -33,6 +33,25 @@ defmodule Nx.Defn.Compiler do
...
+  returns a list of results.
+
+  The callback uses double underscores so it can be defined
+  at root modules without affecting the module's main API.
+  """
+  @callback __compile__(
+              vars :: [Nx.Container.t()],
+              fun :: ([Nx.Container.t()] -> Nx.Container.t()),
+              opts :: keyword
+            ) :: ([[Nx.t()]] -> [Nx.Container.t()])
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/compiler.ex:53]: Can't the input also be `Nx.Container.t()`?

Comment (josevalim) [reply nx/lib/nx/defn/compiler.ex:53]: No, we flatten all containers before.

File: exla/lib/exla/defn.ex [hunk 2]
```diff
@@ -233,7 +234,28 @@ defmodule EXLA.Defn do
...
-  def __jit__(key, vars, fun, [args], options) do
+  def __jit__(key, vars, fun, args_list, options) do
+    {cache_funs, options} = pop_cache_funs(options)
+    __compile__(key, cache_funs, vars, fun, options).(args_list)
+  end
+
+  defp pop_cache_funs(options) do
+    case Keyword.pop(options, EXLA) do
+      {{_, _}, _} = popped ->
+        popped
```
Comment (jonatanklosko) [inline exla/lib/exla/defn.ex:245]: In which case we go this path?

File: nx/lib/nx/defn/compiler.ex (hunk 1 above)
Comment (jonatanklosko) [reply nx/lib/nx/defn/compiler.ex:53]: Makes sense, thanks!

File: exla/lib/exla/defn.ex (hunk 2 above)
Comment (josevalim) [reply exla/lib/exla/defn.ex:245]: `EXLA.jit_cached?` but don't look at it. :D

File: exla/lib/exla/defn.ex [hunk 3]
```diff
@@ -233,7 +234,28 @@ defmodule EXLA.Defn do
...
+
+      {nil, options} ->
+        {{&EXLA.Defn.LockedCache.run/2, &EXLA.Defn.LockedCache.run/2}, options}
+    end
```
Comment (jonatanklosko) [inline exla/lib/exla/defn.ex:249]: ?

```suggestion
    Keyword.pop(options, EXLA, {&EXLA.Defn.LockedCache.run/2, &EXLA.Defn.LockedCache.run/2})
```

**Review by jonatanklosko — APPROVED** [review]

**Pull request merged**